### PR TITLE
coredump: create unnamed temporary files if possible (O_TMPFILE)

### PR DESCRIPTION
**Pull request by evverx**
Don't leave temporary files if the coredump service is aborted during
the operation

> Yeah, these are temporary files that systemd-coredump needs while
> processing the coredumps. Of course, if the coredump service is aborted
> during the operation we better shouldn't leave those files around. This
> is hence a bug to fix in our coredumping code.

See https://github.com/systemd/systemd/issues/2804#issuecomment-210578147

> Another option is to simply use O_TMPFILE, and when it is not available
> fall back to the current behaviour. After all, the files are cleaned up
> eventually, through normal tmpfiles aging, and the offending file
> systems are pretty exotic these days, or not in the upstream kernel.

See https://github.com/systemd/systemd/issues/2804#issuecomment-211496707

### CODE REVIEW

**Comment by evverx** [comment]
How to test:

```
# rm -r /var/lib/systemd/coredump
# systemd-run bash -c "while :; do systemctl stop 'systemd-coredump*.service'; done 2>/dev/null"
# for i in {1..1000}; do bash -c 'sleep 0.1 && kill -SEGV $$'; done
# find /var/lib/systemd/coredump/ -name '.#*'
```

Expected behaviour:

prints nothing on `ext[234]`/`xfs`/`btrfs`/[`f2fs`](https://gist.github.com/evverx/70d903fc3d49fbf42e07f37e60ffb727)
prints something like this

```
...
/var/lib/systemd/coredump/.#core.bash.1000.d78377d90e5346829560c369b0faa2c8.26180.1461034650000000000000f74cf42465a22dd8
...
```

on [`reiserfs`/`zfs`](https://github.com/systemd/systemd/issues/2804#issuecomment-211137127)

File: src/coredump/coredump.c [hunk 1]
```diff
@@ -335,15 +376,11 @@ static int save_external_coredump(
         if (r < 0)
                 return log_error_errno(r, "Failed to determine coredump file name: %m");
 
-        r = tempfn_random(fn, NULL, &tmp);
-        if (r < 0)
-                return log_error_errno(r, "Failed to determine temporary file name: %m");
-
         mkdir_p_label("/var/lib/systemd/coredump", 0755);
```
Comment (evverx) [inline src/coredump/coredump.c:379]: BTW:
(unrelated to this patch, just a reminder)
- there is no `mac_selinux_init`. So, `mkdir_p_label` is the `selinux`-noop.
- `fix_permissions` doesn't call `label_fix`

**Comment by poettering** [comment]
Code looks great, and good to merge, but I wonder if we shouldn't move the two relevant parts into two new calls in src/basic/fileio.c. 

i.e. add a new call fopen_temporary_unlinked() or so, that sits alongside fopen_temporary(), and tries to do the O_TMPFILE thing if possible, and falls back to fopen_temporary() if that doesn't work, exactly as you implemented it. And it should return NULL in the file name pointer if the O_TMPFILE thing worked.  On top of that, add a new call temporary_link() or so, that is the counterpart to this, and links the file into place. It would first do the fflush_and_check, and then either the linkat() or the rename() depending on the temporary filename parameter passed in...

I figure with that in place we could replace quite a few invocations of fopen_temporary() with this new fopen_temporary_unlinked() call (but probably not all), but that'd be material for a later commit.

If you wish we can merge this now, but of course, I think making this generic as suggested above would be even better...

File: src/coredump/coredump.c (hunk 1 above)
Comment (poettering) [inline src/coredump/coredump.c:379]: yeah, this all is a bit chaotic... we often use the mkdir_label() where we shouldn't and mkdir() where we probably should... One of those days we should probably clean this up...

**Comment by evverx** [comment]
> I wonder if we shouldn't move the two relevant parts into two new calls in src/basic/fileio.c

Yeah, we should. But it's not so easy (I've tried) :)

> I think making this generic as suggested above would be even better

I think I'm not ready to properly implement this right now. But I'll take a look.

File: src/coredump/coredump.c [hunk 2]
```diff
@@ -294,6 +308,33 @@ static int make_filename(const char *context[_CONTEXT_MAX], char **ret) {
...
+        if (fd < 0) {
+                log_debug_errno(errno, "Failed to use O_TMPFILE: %m");
+
+                r = tempfn_random(target, NULL, &tmp);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to determine temporary file name: %m");
+
+                fd = open(tmp, O_CREAT|O_EXCL|O_RDWR|O_CLOEXEC|O_NOCTTY|O_NOFOLLOW, 0640);
+                if (fd < 0)
+                        return log_error_errno(errno, "Failed to create coredump file %s: %m", tmp);
```
Comment (fsateler) [inline src/coredump/coredump.c:329]: Why not `unlink(fd)` here? This way, the returned fd has the same behavior always, and no need to check if the name is null in every caller.

Comment (evverx) [inline src/coredump/coredump.c:329]: `unlink(tmp)`, right?
You can't `linkat` that `fd` later:

```
open("hola", O_RDWR|O_CREAT|O_EXCL, 0755) = 3
unlink("hola")                          = 0
linkat(AT_FDCWD, "/proc/self/fd/3", AT_FDCWD, "hola.target", AT_SYMLINK_FOLLOW) = -1 ENOENT (No such file or directory)
```

see `linkat(2)` and `open(2)`

Comment (fsateler) [inline src/coredump/coredump.c:329]: Hmm. Indeed it seems this behavior (or a related one via a new system call) is explicitly forbidden for security reasons 
http://marc.info/?l=linux-kernel&m=104973707408577
http://marc.info/?l=linux-kernel&m=104965552118236&w=2

So the solution was to add a parameter to `open` ... which turned out to be `O_TMPFILE`. OK, sorry for the noise.

**Comment by poettering** [comment]
OK, let's merge this then, we can generalize this in a later commit.

**Comment by poettering** [comment]
@evverx i started generalizing this now, with #3078. But it's only a start, it doesn't deal with fopen_temporary() yet...

**Pull request merged**